core/src/toga/app.py

```diff
@@ -192,137 +243,112 @@ def __init__(
-        windows: Iterable[Window] = (),
         on_exit: OnExitHandler | None = None,
-        factory: None = None,  # DEPRECATED !
     ):
-        """An App is the top level of any GUI program.
-
-        The App is the manager of all the other aspects of execution. An app will
-        usually have a main window; this window will hold the widgets with which the
-        user will interact.
-
-        When you create an App you need to provide a name, an id for uniqueness (by
-        convention, the identifier is a reversed domain name) and an optional startup
-        function which should run once the App has initialized. The startup function
-        constructs the initial user interface. If a startup function is not provided as
-        an argument, you must subclass the App class and define a ``startup()`` method.
-
-        If the name and app_id are *not* provided, the application will attempt to find
-        application metadata. This process will determine the module in which the App
-        class is defined, and look for a ``.dist-info`` file matching that name.
+        """Create a new App instance.
 
-        Once the app is created you should invoke the ``main_loop()`` method, which will
-        start the event loop of your App.
+        Once the app has been created, you should invoke the
+        :meth:`~toga.App.main_loop()` method, which will start the event loop of your
+        App.
 
-        :param formal_name: The formal name of the application. Will be derived from
-            packaging metadata if not provided.
+        :param formal_name: The human-readable name of the app. If not provided,
+            the metadata key ``Formal-Name`` must be present.
         :param app_id: The unique application identifier. This will usually be a
-            reversed domain name, e.g. ``org.beeware.myapp``. Will be derived from
-            packaging metadata if not provided.
-        :param app_name: The name of the Python module containing the app. Will be
-            derived from the module defining the instance of the App class if not
-            provided.
-        :param id: The DOM identifier for the app (optional)
-        :param icon: Identifier for the application's icon.
+            reversed domain name, e.g. ``org.beeware.myapp``. If not provided, the
+            metadata key ``App-ID`` must be present.
+        :param app_name: The name of the distribution used to load metadata with
+            :any:`importlib.metadata`. If not provided, the following will be tried in
+            order:
+
+            #. If there is a ``__main__`` module with a non-empty ``__package__``
+               attribute, that will be used.
+            #. If the ``app_id`` argument was provided, its last segment will be used,
+               with each hyphen replaced with a underscore. For example, an ``app_id``
+               of ``com.example.my-app`` would yield a distribution name of ``my_app``.
```

Comment:
```suggestion
               of ``com.example.my-app`` would yield an ``app_name`` name of ``my_app``.
```

Reply:
Scratch that - I was getting `app_name` and `module_name` turned about in my head. We're not actually trying to import `app_name`, and `importlib_metadata` will accept any distribution name.